packages/hash/frontend/src/components/BlockLoader/shared.ts

```diff
@@ -0,0 +1,112 @@
+import { JsonObject } from "@blockprotocol/core";
+import { BlockGraph, Entity } from "@blockprotocol/graph";
+import get from "lodash/get";
+import set from "lodash/set";
+
+export type SchemaMap = {
+  mapId: string;
+  transformations?: {
+    [destinationKey: string]:
+      | {
+          default?: any;
+          sourceKey?: string;
+        }
+      | undefined;
+  };
+};
+
+type JsonPath = `$.${string}`;
+
+const isJsonPath = (path: unknown): path is JsonPath =>
+  typeof path === "string" && path.startsWith("$.");
+
+/** strips the leading "$." from a jsonpath, if present */
+const pathWithoutLeadingChars = (jsonPath: JsonPath | string) =>
+  jsonPath.replace(/^\$\./, "");
+
+/**
+ * Converts a source entity, linkedEntities and linkGroups into a data tree
+ * @todo support linkedAggregations
+ */
+export const dataTreeFromEntityGraph = (
+  entity: Pick<Entity, "entityId" | "properties">,
+  graphFromEntity: BlockGraph,
+  processedEntityProperties: Map<string, JsonObject> = new Map(),
+) => {
+  const entityPropertiesClone = JSON.parse(JSON.stringify(entity.properties));
```

Comment:
It's possible `immer` may actually be more performant in cases where `entity.properties` is huge. It may also be more readable.

Reply:
There's only one place where we actually mutate this, the `set` below, so it didn't look like it would make it any more readable. I've left as-is until performance becomes a concern